src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,567 @@
+                        if !self.pub_key_set.public_key().verify(sig, &sign_bytes) {
+                            return Err(Error::InvalidMessage(
+                                "invalid hard-vote justification".to_string(),
+                            ));
+                        }
+                    }
+                }
+            }
+            Action::MainVote(action) => {
+                // check the validity of the S-signature share
+                let sign_bytes = self.main_vote_bytes_to_sign(action.round, &action.value)?;
+                if !self
+                    .pub_key_set
+                    .public_key_share(sender)
+                    .verify(&action.sig_share, &sign_bytes)
+                {
+                    return Err(Error::InvalidMessage("invalid signature share".to_string()));
+                }
+
+                match &action.justification {
+                    MainVoteJustification::NoAbstainJustification(sig) => {
+                        // valid S-signature share on the message `(ID, pre-vote, r, b)`
+                        let sign_bytes =
+                            self.pre_vote_bytes_to_sign(action.round, &PreVoteValue::One)?;
+                        if !self.pub_key_set.public_key().verify(sig, &sign_bytes) {
+                            return Err(Error::InvalidMessage(
+                                "invalid hard-vote justification".to_string(),
+                            ));
+                        }
+                    }
+                    MainVoteJustification::AbstainJustification(just_0, just_1) => {
+                        match just_0 {
+                            PreVoteJustification::RoundOneNoJustification => {}
+                            _ => {
+                                return Err(Error::InvalidMessage(format!(
+                                    "invalid justification for value 0 in round 1: {:?}",
+                                    just_0
+                                )));
+                            }
+                        }
+
+                        match just_1 {
+                            PreVoteJustification::RoundOneJustification(subject, proof) => {
+                                if !self
+                                    .pub_key_set
+                                    .public_key()
+                                    .verify(proof, &subject.to_bytes())
+                                {
+                                    return Err(Error::InvalidMessage("invalid proof".to_string()));
+                                }
+                            }
+                            _ => {
+                                return Err(Error::InvalidMessage(format!(
+                                    "invalid justification for value 0 in round 1: {:?}",
+                                    just_1
+                                )));
+                            }
+                        }
```

Comment:
let's add a comment saying that we only expect `abstains` in round 1 therefore we only need to check the `RoundOneJustification`'s